src/libmeasurement_kit/engine.cpp

```diff
@@ -0,0 +1,480 @@
+// Part of Measurement Kit <https://measurement-kit.github.io/>.
+// Measurement Kit is free software under the BSD license. See AUTHORS
+// and LICENSE for more information on the copying conditions.
+
+#define MK_ENGINE_INTERNALS
+#include <measurement_kit/engine.h>
+
+#include <assert.h>
+#include <stdint.h>
+
+#include <algorithm>
+#include <atomic>
+#include <condition_variable>
+#include <deque>
+#include <exception>
+#include <future>
+#include <memory>
+#include <mutex>
+#include <set>
+#include <string>
+#include <sstream>
+#include <thread>
+#include <tuple>
+#include <type_traits>
+
+#include <measurement_kit/common/error.hpp>
+#include <measurement_kit/common/logger.hpp>
+#include <measurement_kit/common/nlohmann/json.hpp>
+#include <measurement_kit/common/reactor.hpp>
+#include <measurement_kit/common/shared_ptr.hpp>
+
+#include "src/libmeasurement_kit/nettests/runnable.hpp"
+
+namespace mk {
+namespace engine {
+
+// # Multi-thread stuff
+//
+// Comes first because it needs more careful handling.
+
+class Semaphore {
+  public:
+    Semaphore() = default;
+
+    void wait() {
+        std::unique_lock<std::mutex> lock{mutex_};
+        cond_.wait(lock, [this]() { return !active_; });
+        active_ = true;
+    }
+
+    void signal() {
+        {
+            std::unique_lock<std::mutex> _{mutex_};
+            active_ = false;
+        }
+        cond_.notify_one();
+    }
+
+    ~Semaphore() = default;
+
+  private:
+    bool active_ = false;
+    std::condition_variable cond_;
+    std::mutex mutex_;
+};
+
+class TaskImpl {
+  public:
+    std::condition_variable cond;
+    std::deque<nlohmann::json> deque;
+    std::atomic_bool interrupted{false};
+    std::mutex mutex;
+    SharedPtr<Reactor> reactor = Reactor::make();
+    std::atomic_bool running{false};
+    std::thread thread;
+};
+
+static void task_run(TaskImpl *pimpl, nlohmann::json &settings);
+static bool is_event_valid(const std::string &);
+
+static void emit(TaskImpl *pimpl, nlohmann::json &&event) {
+
+    // In debug mode, make sure that we're emitting an event that we know
+    assert(event.is_object());
+    assert(event.count("type") != 0);
+    assert(event.at("type").is_string());
+    assert(is_event_valid(event.at("type").get<std::string>()));
+
+    // Actually emit the event.
+    {
+        std::unique_lock<std::mutex> _{pimpl->mutex};
+        pimpl->deque.push_back(std::move(event));
+    }
+    pimpl->cond.notify_all(); // more efficient if unlocked
+}
+
+Task::Task(nlohmann::json &&settings) {
+    pimpl_ = std::make_unique<TaskImpl>();
```

Comment:
Whenever I read this I can't stop myself from thinking of:

![https://www.sanit.org/wp-content/uploads/2018/01/painful-pimple-and-acne.jpg](pimpl)

Reply:
LOL

Reply:
I didn't know that `pimpl` had a meaning in English, I thought it was just an acronym for "pointer to the implementation". I can perhaps `s/pimpl_/impl_/g` as now I find it disgusting.

Reply:
OTOH is a super used C++ idiom, so it makes sense to keep the name in place.